runtime/near-vm-runner/src/cache.rs

```diff
@@ -85,6 +87,102 @@ impl fmt::Debug for MockCompiledContractCache {
     }
 }
 
+#[derive(Clone)]
+pub struct FilesystemCompiledContractCache {
+    dir: Arc<rustix::fd::OwnedFd>,
+}
+
+impl FilesystemCompiledContractCache {
+    pub fn new<HD: AsRef<std::path::Path> + ?Sized, SP: AsRef<std::path::Path> + ?Sized>(
+        home_dir: &HD,
+        store_path: Option<&SP>,
+    ) -> std::io::Result<Self> {
+        let path = std::env::var_os("NEAR_COMPILED_CACHE");
+        let path = path.map(PathBuf::from).unwrap_or_else(|| {
+            let store_path = store_path.map(AsRef::as_ref).unwrap_or_else(|| "data".as_ref());
+            home_dir.as_ref().join(store_path).join("contracts")
+        });
+        std::fs::create_dir_all(&path)?;
+        let dir = rustix::fs::open(path, rustix::fs::OFlags::DIRECTORY, rustix::fs::Mode::empty())?;
+        Ok(Self { dir: Arc::new(dir) })
+    }
+
+    pub fn test() -> std::io::Result<(tempfile::TempDir, Self)> {
```

Comment:
In retrospective, I probably should've made the TempDir a field, rather than making it a responsibility of the caller to figure out how to hold it...